### PR TITLE
Apply Named Entity Recognition

### PR DESCRIPTION
**Pull request by chanteriam**
## Describe your changes
Creates a class for conducting Named Entity Recognition (NER).

## Relevant Issues
#10 

## Checklist before requesting a review
- [x] The code runs successfully.

```commandline
poetry export --without-hashes --format=requirements.txt > requirements.txt
pre-commit run --all-files
trim trailing whitespace.................................................Passed
python tests naming..................................(no files to check)Skipped
check docstring is first.................................................Passed
check that executables have shebangs.................(no files to check)Skipped
check json...........................................(no files to check)Skipped
check for case conflicts.................................................Passed
check toml...............................................................Passed
check for merge conflicts................................................Passed
check xml............................................(no files to check)Skipped
check yaml...............................................................Passed
fix end of files.........................................................Passed
check for broken symlinks............................(no files to check)Skipped
mixed line ending........................................................Passed
sort simple yaml files...............................(no files to check)Skipped
fix python encoding pragma...............................................Passed
pretty format json...................................(no files to check)Skipped
check blanket noqa.......................................................Passed
check for not-real mock methods..........................................Passed
check for eval().........................................................Passed
use logger.warning(......................................................Passed
nbqa-ruff................................................................Passed
nbqa-black...............................................................Passed
nbqa-isort...............................................................Passed
Run isort to sort imports................................................Passed
black....................................................................Passed
ruff.....................................................................Passed
flake8...................................................................Passed
```

### CODE REVIEW

File: legislation_analysis/processing/clean.py [hunk 1]
```diff
@@ -31,8 +35,9 @@ class Cleaner:
         file_name (str): name of file to save.
     """
 
-    DICTIONARY = set(words.words())
+    DICTIONARY = set(words.words()) | MISC_DICTIONARY_ENTRIES
```
Comment (michplunkett) [inline legislation_analysis/processing/clean.py:38]: Gettin' gooooood.

File: legislation_analysis/processing/clean.py [hunk 2]
```diff
@@ -267,15 +300,9 @@ def spell_check(cls, text: str) -> str:
...
-            # check if two words were combined
-            word1, word2 = cls.find_internal_split(word)
-            if word1 and word2:
-                new_words.append(word1)
-                new_words.append(word2)
-                continue
-
-            # misspelled word
-            new_words.append(word)
+            # check if two words were combined or mispelled
```
Comment (michplunkett) [inline legislation_analysis/processing/clean.py:303]: `mispelled` is ... misspelled.

File: legislation_analysis/processing/ner.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+    parameters:
+        file_path (str): path to the file to apply NER to.
+        file_name (str): name of the file to save the NER data to.
+    """
+
+    def __init__(
+        self,
+        file_path,
+        file_name,
```
[the remark this answers is not in the document]
Comment (michplunkett) [reply legislation_analysis/processing/ner.py:37]: Bump.

File: legislation_analysis/processing/ner.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+            "ORG",
+            "GPE",
+            "LAW",
+            "DATE",
+            "EVENT",
+            "CASE",
+        ]
+        new_ner = [ent for ent in new_ner if ent[1] in important_labels]
+
+        return new_ner
```
Comment (michplunkett) [inline legislation_analysis/processing/ner.py:95]: Just return `[ent for ent in new_ner if ent[1] in important_labels]`

File: legislation_analysis/processing/ner.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+                    end -= 1
+
+            # Ensure we don't create an infinite loop
+            if end == start:
+                break
+
+            chunk = text[start:end]
+            ner = cls.apply_ner(chunk, ner)
+
+            start = end  # Move to the next chunk
```
Comment (michplunkett) [inline legislation_analysis/processing/ner.py:135]: Put the comment above the statement.

File: legislation_analysis/processing/ner.py [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+                        aggregated[j] = True  # Mark as aggregated
+
+                # Choose the canonical name as the longest name among similar
+                # entities
+                canonical_name = max(similar_entities, key=len)
+                canonical_entities[(canonical_name, label)] += len(
+                    similar_entities
+                )
+
+        # convert to list of tuples
```
Comment (michplunkett) [inline legislation_analysis/processing/ner.py:209]: You're mixing capitalized and uncapitalized comments.

File: legislation_analysis/processing/ner.py [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+        parameters:
+            cols_to_ner (list): columns to apply NER to.
+        """
+        if cols_to_ner is None:
+            cols_to_ner = [("cleaned_text", "cleaned_text_ner")]
+        self.ner_df = self.df.copy()
+
+        # Apply NER to the text of the legislation
+        for col in cols_to_ner:
+            logging.debug(f"\tApplying NER to {col[0]}...")
```
Comment (michplunkett) [inline legislation_analysis/processing/ner.py:230]: Thoughts on a single period version `...`?

**Review by michplunkett — COMMENTED** [review]
Some small things.

File: legislation_analysis/processing/ner.py (hunk 7 above)
Comment (chanteriam) [reply legislation_analysis/processing/ner.py:231]: i think elipsis is nice

**Comment by michplunkett** [comment]
LGTM!

**Pull request merged**